### PR TITLE
Daef/feature/binary messages

### PR DESCRIPTION
**Pull request by jcheng5**
Resolve merge conflicts and slight tweaks to #1014

### CODE REVIEW

**Comment by jcheng5** [comment]
Don't merge until CLA from daef

**Comment by jcheng5** [comment]
CLA received

File: R/shiny.R [hunk 1]
```diff
@@ -1033,6 +1033,13 @@ ShinySession <- R6Class(
         private$sendResponse(msg, value)
       }
     },
+    sendBinaryMessage = function(type, message) {
+      tagBytes <- charToRaw(type)
```
Comment (wch) [inline R/shiny.R]: Should this be `typeBytes` and the error message below be "Type too long"?

File: srcjs/shinyapp.js [hunk 2]
```diff
@@ -487,14 +500,14 @@ var ShinyApp = function() {
     this.$updateConditionals();
   };
 
-
   // A function for sending messages to the appropriate handlers.
   // - msgObj: the object containing messages, with format {msgObj.foo, msObj.bar
   this._sendMessagesToHandlers = function(msgObj, handlers, handlerOrder) {
+    var tag, buffer, i;
```
Comment (wch) [inline srcjs/shinyapp.js]: The `tag` and `buffer` vars don't appear to be used -- maybe this should just be left unchanged.

File: srcjs/shinyapp.js [hunk 3]
```diff
@@ -473,8 +472,22 @@ var ShinyApp = function() {
...
 
-  this.dispatchMessage = function(msg) {
-    var msgObj = JSON.parse(msg);
+  this.dispatchMessage = function(data) {
+    var msgObj = {};
+    if(typeof data === "string") {
+      msgObj = JSON.parse(data);
+    } else { // data is arraybuffer
+      var len = new DataView(data,0,1).getUint8(0);
+      var tagdv = new DataView(data,1,len);
```
Comment (wch) [inline srcjs/shinyapp.js]: It would be helpful to rename "tag" here to "type", for consistency with the R code.

**Pull request merged**